File: LeetCode/Easy/0202-happy-number/0202-happy-number.go
func isHappy(n int) bool {
    m := make(map[int]bool)
    
    for n != 1 {
        exist, val := m[n]
        if (exist && val) {
            return false
        }
        m[n] = true
        x := 0
        for n != 0 {
            x += (n % 10) * (n % 10)
            n = n / 10
        }
        n = x
    }
    return true
}

File: LeetCode/Easy/0202-happy-number/0202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> visit = new HashSet<>();

        while (true) {
            if (n == 1) {
                break;
            }
            if (visit.contains(n)) {
                return false;
            }
            visit.add(n);
            n = replace(n);
        }
        return true;
    }

    public int replace(int x) {
        int ret = 0;
        while (x != 0) {
            ret += (x % 10) * (x % 10);
            x = x / 10;
        }
        return ret;
    }
}